File: src/0695-Max-Area-of-Island/0695.go
var r, c int
var d []int

func maxAreaOfIsland(grid [][]int) int {
    r, c = len(grid), len(grid[0])
    d = []int{0, 1, 0, -1, 0}
    res := 0

    for i := 0; i < r; i++ {
        for j := 0; j < c; j++ {
            if grid[i][j] == 1 {
                grid[i][j] = 0
                res = max(res, dfs(grid, i, j))
            }
        }
    }
    return res
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func dfs(grid [][]int, x, y int) int {
    res := 1
    for i := 0; i < 4; i++ {
        nx, ny := x + d[i], y + d[i + 1]
        if 0 <= nx && nx < r && 0 <= ny && ny < c && grid[nx][ny] == 1 {
            grid[nx][ny] = 0
            res += dfs(grid, nx, ny)
        }
    }
    return res
}

File: src/0695-Max-Area-of-Island/0695.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        r, c = len(grid), len(grid[0])
        d = [0, 1, 0, -1, 0]
        
        def dfs(x, y):
            res = 1
            for i in range(4):
                nx, ny = d[i] + x, d[i + 1] + y
                if 0 <= nx < r and 0 <= ny < c and grid[nx][ny] == 1:
                    grid[nx][ny] = 0
                    res += dfs(nx, ny)
            return res

        res = 0
        for i in range(r):
            for j in range(c):
                if grid[i][j] == 1:
                    grid[i][j] = 0
                    res = max(res, dfs(i, j))
        return res